pkg/controller/controller.go

```diff
@@ -386,11 +395,10 @@ func (c *Controller) handleActionSet(as *crv1alpha1.ActionSet) (err error) {
 	}()
 
 	for i := range as.Status.Actions {
-		if err = c.runAction(ctx, as, i); err != nil {
+		bp := bps.Get(as.Spec.Actions[i].Blueprint)
```

Comment:
What if bp == nil? (not found in the cache)

Reply:
It's checked that bp is in cache during init action set stage, so no need to check it here one more time

Reply:
Cannot it happen that between you init the cache and you make a call to bps.Get something will change? In other words, do we have any kind of lock that guarantees the atomicity of the operations?

Reply:
Currently, all information required for phase execution is stored inside of Blueprint object. With introducing bp cache the code became more atomic than it was before, because each onAdd has a snapshot of blueprint information required for phases. Unfortunately, the whole code is not atomic, but this is out of the task scope.

Reply:
Moved bp cache initialization function to controller

Reply:
> Unfortunately, the whole code is still not atomic, but this is out of the task scope.

I think, if you just check the bp value (ok in your current approach the value of `ok`), then it will be safer than just hope.

Reply:
Ok, will add this check here. But the cache is not shared among onAdd event handlers, so nothing will happen with it after initialization.

Reply:
Fixed